rust/src/operations/vacuum.rs

```diff
@@ -0,0 +1,325 @@
+            {
+                continue;
+            }
+
+            files_to_delete.push(obj_meta.location);
+        }
+
+        Ok(VacuumPlan { files_to_delete })
+    }
+}
+
+impl std::future::IntoFuture for VacuumBuilder {
+    type Output = DeltaResult<(DeltaTable, VacuumMetrics)>;
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let plan = this.create_vacuum_plan().await?;
+            if this.dry_run {
+                return Ok((
+                    DeltaTable::new_with_state(this.store, this.snapshot),
+                    VacuumMetrics {
+                        files_deleted: plan.files_to_delete.iter().map(|f| f.to_string()).collect(),
+                        dry_run: true,
+                    },
+                ));
+            }
+
+            let metrics = plan.execute(&this.store).await?;
+            Ok((
+                DeltaTable::new_with_state(this.store, this.snapshot),
+                metrics,
+            ))
+        })
+    }
+}
+
+/// Encapsulate which files are to be deleted and the parameters used to make that decision
+struct VacuumPlan {
+    /// What files are to be deleted
+    pub files_to_delete: Vec<Path>,
+}
+
+impl VacuumPlan {
+    /// Execute the vacuum plan and delete files from underlying storage
+    pub async fn execute(self, store: &DeltaObjectStore) -> Result<VacuumMetrics, DeltaTableError> {
+        if self.files_to_delete.is_empty() {
+            return Ok(VacuumMetrics {
+                dry_run: false,
+                files_deleted: Vec::new(),
+            });
+        }
+
+        // Delete the files
+        let files_deleted = match store.delete_batch(&self.files_to_delete).await {
+            Ok(_) => Ok(self.files_to_delete),
+            Err(err) => Err(err),
+        }?
+        .into_iter()
+        .map(|file| file.to_string())
+        .collect();
+
+        Ok(VacuumMetrics {
+            files_deleted,
+            dry_run: false,
+        })
+    }
+}
+
+/// Whether a path should be hidden for delta-related file operations, such as Vacuum.
+/// Names of the form partitionCol=[value] are partition directories, and should be
+/// deleted even if they'd normally be hidden. The _db_index directory contains (bloom filter)
+/// indexes and these must be deleted when the data they are tied to is deleted.
+fn is_hidden_directory(snapshot: &DeltaTableState, path: &Path) -> Result<bool, DeltaTableError> {
+    let path_name = path.to_string();
+    Ok((path_name.starts_with('.') || path_name.starts_with('_'))
+        && !path_name.starts_with("_delta_index")
+        && !path_name.starts_with("_change_data")
+        && !snapshot
+            .current_metadata()
+            .ok_or(DeltaTableError::NoMetadata)?
+            .partition_columns
+            .iter()
+            .any(|partition_column| path_name.starts_with(partition_column)))
```

Comment:
Just deleting the line unfortunately led to failing tests, but I did not dive deeper. My general feeling is we have to re-visit the logic anyhow, since we are not yet cleaning up "associated" files. I can look further into this here, but would prefer in a follow up :) - maybe when doing #688.

Reply:
Okay. Thanks for looking into it.